client_configuration_finder.go

```diff
@@ -0,0 +1,71 @@
+// Copyright (c) 2023 Snowflake Computing Inc. All rights reserved.
+
+package gosnowflake
+
+import (
+	"errors"
+	"os"
+	"path"
+)
+
+const (
+	defaultConfigName = "sf_client_config.json"
+	clientConfEnvName = "SF_CLIENT_CONFIG_FILE"
+)
+
+func getClientConfig(filePathFromConnectionString string) (*ClientConfig, error) {
+	configPredefinedFilePaths, err := clientConfigPredefinedDirs()
+	var filePath string
+	filePath, err = findClientConfig(filePathFromConnectionString, configPredefinedFilePaths)
+	if err != nil {
+		return nil, err
+	}
+	if filePath == "" { // we did not find a config file
+		return nil, nil
+	}
+	return parseClientConfiguration(filePath)
+}
+
+func findClientConfig(filePathFromConnectionString string, configPredefinedDirs []string) (string, error) {
+	if filePathFromConnectionString != "" {
+		return filePathFromConnectionString, nil
+	}
+	envConfigFilePath := os.Getenv(clientConfEnvName)
+	if envConfigFilePath != "" {
+		return envConfigFilePath, nil
+	}
+	return searchForConfigFile(configPredefinedDirs)
+}
+
+func searchForConfigFile(directories []string) (string, error) {
+	for _, dir := range directories {
+		filePath := path.Join(dir, defaultConfigName)
+		exists, err := existsFile(filePath)
+		if err != nil {
+			return "", err
+		}
+		if exists {
+			return filePath, nil
+		}
+	}
+	return "", nil
+}
+
+func existsFile(filePath string) (bool, error) {
+	_, err := os.Stat(filePath)
+	if err == nil {
+		return true, nil
+	}
+	if errors.Is(err, os.ErrNotExist) {
+		return false, nil
+	}
+	return false, err
```

Comment:
in the previous version, when we mocked existsFile operation I was easyily able to mock the case of returning error other than ErrNotExist. How would you test it? I tried to write a test by specifying file permissions 0000 or creating a directory instead of file, but I'm not able to get error other than ErrNotExist.

Reply:
Ok, I checked that we have something like that (source: os/error.go):
```
ErrPermission = fs.ErrPermission // "permission denied"
```
But if permissions without read not works so we can skip it.